lib/maths/CTreeShapFeatureImportance.cc

```diff
@@ -200,67 +213,87 @@ void CTreeShapFeatureImportance::extendPath(SPath& path,
     // to sets of size i and we **also** need to scale by the difference in binomial coefficients as both M
     // increases by one and i increases by one. So we get additive term 1{last feature selects path if in S}
     // * scale(i) * (i+1)! (M+1-(i+1)-1)!/(M+1)! / (i! (M-i-1)!/ M!), whence += scale(i) * (i+1) / (M+1).
-    path.extend(featureIndex, fractionZero, fractionOne);
-    int pathDepth = static_cast<int>(path.depth());
-    for (int i = (pathDepth - 1); i >= 0; --i) {
-        path.s_Scale[i + 1] += fractionOne * path.s_Scale[i] *
-                               static_cast<double>(i + 1.0) /
-                               static_cast<double>(pathDepth + 1.0);
-        path.s_Scale[i] = fractionZero * path.s_Scale[i] *
-                          static_cast<double>(pathDepth - i) /
-                          static_cast<double>(pathDepth + 1.0);
+
+    path.setValues(nextIndex, fractionOne, fractionZero, featureIndex);
+    if (nextIndex == 0) {
+        scalePath[nextIndex] = 1.0;
+    } else {
+        scalePath[nextIndex] = 0.0;
+    }
+    double countDown{static_cast<double>(nextIndex)};
+    double countUp{1.0};
+    for (int i = (nextIndex - 1); i >= 0; --i, --countDown, ++countUp) {
+        scalePath[i + 1] += scalePath[i] * countDown * fractionOne /
+                            static_cast<double>(nextIndex + 1);
+        scalePath[i] *= countUp * fractionZero / static_cast<double>(nextIndex + 1);
```

Comment:
How about:
```
double stepDown{fractionOne /  static_cast<double>(nextIndex + 1)};
double stepUp{fractionZero /  static_cast<double>(nextIndex + 1)};
double countDown{static_cast<double>(nextIndex) * stepDown};
double countUp{stepUp};
for (int i = (nextIndex - 1); i >= 0; --i, countDown -= stepDown, countUp += stepUp) {
  scalePath[i + 1] += scalePath[i] * countDown;
  scalePath[i] *= countUp;
}
```
This saves two multiplications and divisions per iteration. (I don't think the compiler could make these rewrites because the rounding errors would be different.)

Reply:
I was experimenting with this and didn't see much of a difference with -O3. But this makes the code much more readible, so I changed it in cbd15587